tools/closed-by-commits

```diff
@@ -0,0 +1,436 @@
+                    }
+                  }
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+    """
+
+    def __init__(self, token: str, reponame: str, czo_issue_prefix: str) -> None:
+        self.github = Github(auth=Auth.Token(token))
+        self.reponame = reponame
+        self.czo_issue_prefix = czo_issue_prefix
+
+    @staticmethod
+    def _extract_czo_urls(text: str | None) -> set[str]:
+        if not text:
+            return set()
+
+        matches = re.findall(r"https://chat\.zulip\.org/[^\s\)\]\>]+", text)
+
+        urls = set()
+        for url in matches:
+            # We strip off and remove /with/... and /near/... to
+            # reduce the number of unique links which are generated.
+            parsed_url = re.match(r"(.*)/topic/([^/]+)(/(near|with)/.*)?$", url)
+            if not parsed_url:
+                continue
+            urls.add(
+                parsed_url[1]
+                + "/topic/"
+                # Normalize the topic by decoding and re-encoding
+                + encode_hash_component(urllib.parse.unquote(parsed_url[2].replace(".", "%")))
+            )
+
+        return urls
+
+    @staticmethod
+    def _extract_issue_numbers(reponame: str, text: str | None) -> set[int]:
+        if not text:
+            return set()
+
+        # https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword
+        pattern = rf"\b(?:close[sd]?|fix(?:es|ed)?|resolve[sd]?):?\s+(?:zulip/{reponame})?#(\d+)\b"
+        matches = re.findall(pattern, text, re.IGNORECASE)
+        return {int(num) for num in matches}
+
+    def _issue_from_node(self, issue_node: dict[str, Any]) -> Issue:
+        issue = Issue(
+            number=issue_node["number"],
+            title=issue_node["title"].strip(),
+        )
+        issue_comments = issue_node.get("comments", {}).get("nodes", [])
+        for node in [issue_node, *issue_comments]:
+            issue.czo_urls.update(self._extract_czo_urls(node.get("body")))
+        issue.czo_urls.update(search_czo_for_number(self.czo_issue_prefix, issue.number))
+
+        issue.duplicate_issue_ids = {
+            event["duplicate"]["number"] for event in issue_node["timelineItems"]["nodes"]
+        }
+        for issue_id in issue.duplicate_issue_ids:
+            if duplicate_info := self._fetch_issue(issue_id):
+                issue.czo_urls.update(duplicate_info.czo_urls)
+                issue.duplicate_issue_ids.update(duplicate_info.duplicate_issue_ids)
+        return issue
+
+    def _fetch_issue(self, number: int) -> Issue | None:
+        """Fetch issue metadata from GitHub."""
+        try:
+            _, result = self.github.requester.graphql_query(
+                self.ISSUE_QUERY, {"number": number, "repo": self.reponame}
+            )
+
+            if "errors" in result:
+                return None
```

Comment:
I guess does this mean we'll just abort if GitHub was unstable at the time? Seems OK especially if we can get the right command to rerun manually from the CI logs.

Though do we want to print something more?